### PR TITLE
Improve logging

### LINKED ISSUE #17: Improve logging
Appropriate log messages should be used to help debug or automate testing through travis/jenkins.
Also add a debug mode.

### PR DESCRIPTION
**Pull request by ShishirH**
Closes #17 

1. Added log4j implementation of slf4j.
2. Added a log4j2.properties file for logging configuration.
3. Added a new env LOGGING_LEVEL that can be set in the YAML file.

Signed-off-by: Shishir Halaharvi <shihala1@in.ibm.com>

### CODE REVIEW

**Comment by ShishirH** [comment]
@dinogun Should be held off until minikube PR is merged and additional changes are made.

**Comment by ShishirH** [comment]
@chandrams Do you have any suggestions on the log messages you need for testing?

**Comment by ShishirH** [comment]
@dinogun @chandrams Should the logs be saved in a file, or should they be directed to STDOUT?

They would be persistent in a file, but anyone who wants to get that log file would need to run `kubectl cp`, while the logs can be easily viewed by using `kubectl logs -f` for STDOUT.
Should this be an option for the user?

**Comment by dinogun** [comment]
The k8s best practice is stdout, no files please.

**Comment by ShishirH** [comment]
For reference:
1. When DEBUG_LEVEL is set to INFO, or no env is set(by default) [INFO_LOG.txt](https://github.com/kruize/kruize/files/3890846/INFO_LOG.txt)

2. When DEBUG_LEVEL is set to DEBUG: [DEBUG_LOG.txt](https://github.com/kruize/kruize/files/3890847/DEBUG_LOG.txt)

**Comment by chandrams** [comment]
1) Info provided in info log is more, we shouldn't be printing recommendations for idle apps. We can have a logic to monitor only active applications. Have a thread monitoring the idle ones and move them to active when they become active.

2) Do we need to log the Auth token and Monitoring agent end point? Would there be any security issue? We can anyways print this when Debug is on.

3) Memory is in KB or MB? Can we print this too.

4) Can we print the original requests and limits when INFO is on?

5) Can the sampling interval be configurable? I think this can be done if kruize monitors apps in parallel. 

6) In the info log, it prints "No previous recommendations available" - include the app name here.

7) If an app is scaled, does kruize give recommendations for all scaled applications. We need to test this scenario.

8) If a pod is removed, does kruize stop monitoring and logging this.

9) What error messages are logged when an error occurs? Does this msg "com.kruize.main.Kruize - End of initialization phase" indicate kruize has been successfully initialized.

**Comment by chandrams** [comment]
10. Should we log recommendations each time kruize monitors and generates recommendations? Can we have it log every 5th or 'X'th time it generated the recommendations. This would reduce the size of the log. Atleast in INFO level we don't need this, might be DEBUG can give these details if required.

**Comment by ShishirH** [comment]
>     1. Info provided in info log is more, we shouldn't be printing recommendations for idle apps. We can have a logic to monitor only active applications. Have a thread monitoring the idle ones and move them to active when they become active.

Created an issue #58. But this will take some time. 

>     2. Do we need to log the Auth token and Monitoring agent end point? Would there be any security issue? We can anyways print this when Debug is on.

True, security would be a major concern. I think that Auth token shouldn't be printed even in debug mode, but a little conflicted about the endpoint URL. Makes debugging easy, but still poses a security risk. I would welcome your thoughts on how to approach this @chandrams @dinogun 

>     3. Memory is in KB or MB? Can we print this too.

Will make the change.

>     4. Can we print the original requests and limits when INFO is on?

Will make the change .

>     5. Can the sampling interval be configurable? I think this can be done if kruize monitors apps in parallel.

Should be done in #58, but again, looks like it'll take quite a while until that's done.

>     6. In the info log, it prints "No previous recommendations available" - include the app name here.

Will make the change.

>     7. If an app is scaled, does kruize give recommendations for all scaled applications. We need to test this scenario.

Will test and print the logs.
 
>     8. If a pod is removed, does kruize stop monitoring and logging this.

Will test and print the logs.

>     9. What error messages are logged when an error occurs? Does this msg "com.kruize.main.Kruize - End of initialization phase" indicate kruize has been successfully initialized.

Will give the logs when initialization fails.

**Comment by ShishirH** [comment]
Removed printing auth token.
Printing memory in MB now.
Bug with original values results in them being obtained multiple times as of now. This should be resolved once #44 gets solved, following which original values will also be made available in INFO.

**Comment by ShishirH** [comment]
Errors when initialization fails:
[initializationError3.txt](https://github.com/kruize/kruize/files/3915033/initializationError3.txt)
[initializationError2.txt](https://github.com/kruize/kruize/files/3915034/initializationError2.txt)
[initializationError.txt](https://github.com/kruize/kruize/files/3915036/initializationError.txt)

**Comment by ShishirH** [comment]
Logs when acmeair pod was deleted after sometime:
[podDeleteInfo.txt](https://github.com/kruize/kruize/files/3915392/podDeleteInfo.txt)
[deletePodDebug.txt](https://github.com/kruize/kruize/files/3915398/deletePodDebug.txt)


No errors were printed or additional messages were printed in this case and the same earlier recommendations were shown. Do you want anything specific in the logs for this case showing that the app was killed?

File: src/main/java/com/kruize/environment/docker/DockerEnvImpl.java [hunk 1]
```diff
@@ -16,21 +16,25 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.FileReader;
 import java.io.IOException;
 import java.util.ArrayList;
 
 public class DockerEnvImpl extends EnvTypeImpl
 {
+    private static Logger logger = LoggerFactory.getLogger(DockerEnvImpl.class);
```
Comment (chandrams) [inline src/main/java/com/kruize/environment/docker/DockerEnvImpl.java]: logger variable name can be consistent and as per coding guidelines

**Comment by chandrams** [comment]
Logs look fine. Should we reduce the frequency for logging recommendations for a deleted app?

File: src/main/java/com/kruize/environment/docker/DockerEnvImpl.java (hunk 1 above)
Comment (ShishirH) [reply src/main/java/com/kruize/environment/docker/DockerEnvImpl.java]: Made all loggers final, and variable name to uppercase

**Comment by ShishirH** [comment]
@chandrams Yes, deleted apps shouldn't log as frequently. But right now, there is no convenient way of checking if the application was killed. Will need to add that in a future PR.

**Comment by chandrams** [comment]
@dinogun - I have this PR, few changes requested will need further work. For now, I think we can go ahead and merge the changes.

**Comment by dinogun** [comment]
Can we make this small in the logs
```
Timestamp  LOG_TYPE FUNCTION_NAME pod_name: ACTIVITY_INFO
```
This would translate as follows
```
07:16:43.653 [Thread-4] RECO  com.kruize.analysis.AnalysisImpl - acmeair-sample-57c4bdc97-nq7pn: CPU Limit = 1.2
07:16:43.653 [Thread-4] RECO  com.kruize.analysis.AnalysisImpl - acmeair-sample-57c4bdc97-nq7pn: CPU Requests = 1.1
07:16:43.654 [Thread-4] RECO  com.kruize.analysis.AnalysisImpl - acmeair-sample-57c4bdc97-nq7pn: Memory Limit = 222.12 MB
07:16:43.654 [Thread-4] RECO  com.kruize.analysis.AnalysisImpl - acmeair-sample-57c4bdc97-nq7pn: Memory Requests = 172.12109565734863 MB
```
Makes it a lot more easier to read. Also note `RECO` tag instead of `INFO` for recommendations. This can be a different log level, so that should make it easy to turn on just the recommendations for debugging if needed.

**Comment by ShishirH** [comment]
@dinogun Changed the log messages as in the above comment. Currently though, the custom RECO tag is not possible because slf4j doesn't support it. Right now, as we are using Eclipse Jetty, which depends on slf4j, it is a little difficult to migrate to other logging libraries like logback or log4j. 

In the future, as Kruize will need an API endpoint, we should use OpenLiberty or other better webservers rather than Jetty. That would enable us to completely switch to log4j, and use custom tags.

File: src/main/resources/log4j2.properties [hunk 2]
```diff
@@ -0,0 +1,12 @@
+appenders=kruize
+
+appender.kruize.type = Console
+appender.kruize.name = myOutput
```
Comment (dinogun) [inline src/main/resources/log4j2.properties]: Can this be more descriptive ? What exactly does `myOutput` signify ?

Comment (ShishirH) [reply src/main/resources/log4j2.properties]: Its the name of the appender for the logger to get a reference. Changed its name to kruizeAppender.

**Comment by ShishirH** [comment]
@dinogun Updated README.

**Comment by ShishirH** [comment]
Also changed the env from `DEBUG_LEVEL` to `LOGGING_LEVEL`. Is this OK or should I roll it back?

**Pull request merged**